lib/base/json.cpp

```diff
@@ -2,208 +2,354 @@
+			break;
+		case ValueNumber:
+			EncodeStringOrNumber(value.Get<double>());
+			break;
+		case ValueObject: {
+			const auto& obj = value.Get<Object::Ptr>();
+			const auto& type = obj->GetReflectionType();
+			if (type == Namespace::TypeInstance) {
+				static constexpr auto extractor = [](const NamespaceValue& v) -> const Value& { return v.Val; };
+				EncodeObject(static_pointer_cast<Namespace>(obj), extractor, yc);
+			} else if (type == Dictionary::TypeInstance) {
+				static constexpr auto extractor = [](const Value& v) -> const Value& { return v; };
+				EncodeObject(static_pointer_cast<Dictionary>(obj), extractor, yc);
+			} else if (type == Array::TypeInstance) {
+				EncodeArray(static_pointer_cast<Array>(obj), yc);
+			} else if (auto gen(dynamic_pointer_cast<ValueGenerator>(obj)); gen) {
+				EncodeValueGenerator(gen, yc);
+			} else {
+				// Some other non-serializable object type!
+				EncodeStringOrNumber(obj->ToString());
+			}
 			break;
+		}
+		default:
+			VERIFY(!"Invalid variant type.");
+	}
+
+	// If we are at the top level of the JSON object and prettifying is enabled, we need to end
+	// the JSON with a newline character to ensure that the output is properly formatted.
+	if (m_Indent == 0 && m_Pretty) {
+		Write("\n");
+	}
+}
 
-		case ValueObject:
-			{
-				const Object::Ptr& obj = value.Get<Object::Ptr>();
+/**
+ * Encodes an Array object into JSON and writes it to the output stream.
+ *
+ * @param array The Array object to be serialized into JSON.
+ * @param yc The optional yield context for asynchronous operations. If provided, it allows the encoder
+ * to flush the output stream safely when it has not acquired any object lock.
+ */
+void JsonEncoder::EncodeArray(const Array::Ptr& array, boost::asio::yield_context* yc)
+{
+	BeginContainer('[');
+	auto olock = array->LockIfRequired();
+	if (olock) {
+		yc = nullptr; // We've acquired an object lock, never allow asynchronous operations.
+	}
 
-				{
-					Namespace::Ptr ns = dynamic_pointer_cast<Namespace>(obj);
-					if (ns) {
-						EncodeNamespace(stateMachine, ns);
-						break;
-					}
-				}
+	bool isFirst = true;
+	for (const auto& item : array) {
+		WriteSeparatorAndIndentStrIfNeeded(!isFirst);
+		isFirst = false;
+		Encode(item, yc);
+		FlushIfSafe(yc);
+	}
+	EndContainer(']', isFirst);
+}
+
+/**
+ * Encodes a ValueGenerator object into JSON and writes it to the output stream.
+ *
+ * This will iterate through the generator, encoding each value it produces until it is exhausted.
+ *
+ * @param generator The ValueGenerator object to be serialized into JSON.
+ * @param yc The optional yield context for asynchronous operations. If provided, it allows the encoder
+ * to flush the output stream safely when it has not acquired any object lock on the parent containers.
+ */
+void JsonEncoder::EncodeValueGenerator(const ValueGenerator::Ptr& generator, boost::asio::yield_context* yc)
+{
+	BeginContainer('[');
+	bool isEmpty = true;
```

Comment:
If you prefer `isFirst` over `isEmpty`, please use it here as well. The functionality is exactly the same for array/dictionary/generator, so the name should be as well.

Reply:
Done!